distributed/scheduler.py

```diff
@@ -4082,6 +4084,105 @@ def _to_dict(
+        finally:
+            # Ensure the tasks aren't left running if anything fails.
+            # Someday (py3.11), use a trio-style TaskGroup for this.
+            workers_future.cancel()
+
+        # Convert any RPC errors to strings
+        worker_states = {
+            k: repr(v) if isinstance(v, Exception) else v
+            for k, v in worker_states.items()
+        }
+
+        return {
+            "scheduler": scheduler_state,
+            "workers": worker_states,
+            "versions": {"scheduler": self.versions(), "workers": worker_versions},
+        }
+
+    async def dump_cluster_state_to_url(
+        self,
+        url: str,
+        exclude: Collection[str],
+        format: Literal["msgpack", "yaml"],
+        storage_options: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        "Write a cluster state dump to an fsspec-compatible URL. Requires fsspec."
+        try:
+            import fsspec
+        except ImportError:
+            raise ImportError(
+                "fsspec must be installed to dump cluster state directly to a URL"
+            )
+
+        if storage_options is None:
+            storage_options = {}
+
+        if format == "msgpack":
+            import msgpack
+
+            # NOTE: `compression="infer"` will automatically use gzip via the `.gz` suffix
+            mode = "wb"
+            suffix = ".msgpack.gz"
+            if not url.endswith(suffix):
+                url += suffix
+            writer = msgpack.pack
+        elif format == "yaml":
+            import yaml
+
+            mode = "w"
+            suffix = ".yaml"
+            if not url.endswith(suffix):
+                url += suffix
+
+            def writer(state: dict, f):
+                # YAML adds unnecessary `!!python/tuple` tags; convert tuples to lists to avoid them.
+                def tuple_to_list(node):
+                    if isinstance(node, (list, tuple)):
+                        return [tuple_to_list(el) for el in node]
+                    elif isinstance(node, dict):
+                        return {k: tuple_to_list(v) for k, v in node.items()}
+                    else:
+                        return node
+
+                state = tuple_to_list(state)
+                yaml.dump(state, f)
+
+        else:
+            raise ValueError(
+                f"Unsupported format {format!r}. Possible values are 'msgpack' or 'yaml'."
+            )
+
+        # Eagerly open the file to catch any errors before doing the full dump
+        with fsspec.open(url, mode, compression="infer", **storage_options) as f:
+            state = await self.get_cluster_state(exclude)
+
+            # Write from a thread so we don't block the event loop quite as badly
+            # (the writer will still hold the GIL a lot though).
+            # TODO use `asyncio.to_thread` once <3.9 support is dropped.
+            await self.loop.run_in_executor(None, writer, state, f)
```

Comment:
I assume that serializing and writing the dict should be safe to do from another thread as long; does anyone see a reason why not?

Reply:
afaik, this should be safe. accessing is fine as long as nothing is mutated